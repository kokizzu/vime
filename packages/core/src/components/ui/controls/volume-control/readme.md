# vime-volume-control

A control for adjusting the volume of the player and toggling the muted state.

## Example

```html
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <vime-volume-control></vime-volume-control>
    </vime-controls>
  </vime-ui>
<vime-player>
```

<!-- Auto Generated Below -->


## Properties

| Property           | Attribute            | Description                                                                                                                                                             | Type                             | Default               |
| ------------------ | -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------- | --------------------- |
| `hideTooltip`      | `hide-tooltip`       | Whether the tooltip should be hidden.                                                                                                                                   | `boolean`                        | `false`               |
| `highVolumeIcon`   | `high-volume-icon`   | The URL to an SVG element or fragment.                                                                                                                                  | `string`                         | `'#vime-volume-high'` |
| `lowVolumeIcon`    | `low-volume-icon`    | The URL to an SVG element or fragment.                                                                                                                                  | `string`                         | `'#vime-volume-low'`  |
| `muteKeyCodes`     | `mute-key-codes`     | A pipe (`\|`) seperated string of JS key codes, that when caught in a `keydown` event, will toggle the muted state of the player.                                       | `string \| undefined`            | `'77'`                |
| `muteKeyboardHint` | `mute-keyboard-hint` | If the `keyCodes` prop is provided, this prop can provide a hint to the user inside the control tooltip, on what key presses will toggle the muted state of the player. | `string \| undefined`            | `'(m)'`               |
| `mutedIcon`        | `muted-icon`         | The URL to an SVG element or fragment.                                                                                                                                  | `string`                         | `'#vime-volume-mute'` |
| `noKeyboard`       | `no-keyboard`        | Prevents the volume being changed using the Up/Down arrow keys.                                                                                                         | `boolean`                        | `false`               |
| `tooltipDirection` | `tooltip-direction`  | The direction in which the tooltip should grow.                                                                                                                         | `"left" \| "right" \| undefined` | `undefined`           |


## Dependencies

### Used by

 - [vime-default-controls](../default-controls)

### Depends on

- [vime-mute-control](../mute-control)
- [vime-slider](../../slider)

### Graph
```mermaid
graph TD;
  vime-volume-control --> vime-mute-control
  vime-volume-control --> vime-slider
  vime-mute-control --> vime-control
  vime-mute-control --> vime-icon
  vime-mute-control --> vime-tooltip
  vime-default-controls --> vime-volume-control
  style vime-volume-control fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
