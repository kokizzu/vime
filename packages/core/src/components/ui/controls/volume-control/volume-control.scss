vime-volume-control {
  @function slider-transition($prop) {
    @return $prop 0.2s cubic-bezier(0.4, 0, 1, 1);
  }

  align-items: center;
  display: flex;
  position: relative;
  pointer-events: auto;
  margin-left: calc(#{$control-spacing} / 2) !important;

  vime-slider {
    width: 75px;
    height: 100%;
    margin: 0;
    max-width: 0;
    position: relative;
    z-index: 3;
    transition: slider-transition(margin), slider-transition(max-width);
    margin-left: calc(#{$control-spacing} / 2) !important;
    visibility: hidden;

    &:hover {
      cursor: pointer;
    }

    &.hidden {
      display: none;
    }

    &.active {
      max-width: 75px;
      margin: 0 ($control-spacing / 2);
      visibility: visible;
    }
  }
}