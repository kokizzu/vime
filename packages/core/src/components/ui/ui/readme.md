# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop
is `true`, and the player is not in fullscreen mode.

<!-- Auto Generated Below -->

## Usage

### Angular

```html {3-5} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### Html

```html {3-5}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### React

```tsx {2,8}
import React from 'react';
import { VimePlayer, VimeUi } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>{/* ... */}</VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {4-6,11,16} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
    },
  };
</script>
```

## Dependencies

### Used by

- [vime-default-ui](../default-ui)

### Graph

```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
