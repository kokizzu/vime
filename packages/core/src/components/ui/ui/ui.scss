vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}