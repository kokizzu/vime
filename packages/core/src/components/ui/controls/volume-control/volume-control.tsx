import {
  h, Host, Component, Prop, State, Watch,
} from '@stencil/core';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';
import { PlayerStateDispatcher, createPlayerStateDispatcher } from '../../../core/player/PlayerState';
import { TooltipDirection } from '../../tooltip/types';
import { Disposal } from '../../../core/player/Disposal';
import { listen } from '../../../../utils/dom';
import { findRootPlayer } from '../../../core/player/utils';

@Component({
  tag: 'vime-volume-control',
  styleUrl: 'volume-control.scss',
})
export class VolumeControl {
  private dispatch!: PlayerStateDispatcher;

  private keyboardDisposal = new Disposal();

  private prevMuted = false;

  private hideSliderTimeout?: number;

  @State() currentVolume = 50;

  @State() isSliderActive = false;

  /**
   * The URL to an SVG element or fragment.
   */
  @Prop() lowVolumeIcon = '#vime-volume-low';

  /**
   * The URL to an SVG element or fragment.
   */
  @Prop() highVolumeIcon = '#vime-volume-high';

  /**
   * The URL to an SVG element or fragment.
   */
  @Prop() mutedIcon = '#vime-volume-mute';

  /**
   * The direction in which the tooltip should grow.
   */
  @Prop() tooltipDirection: TooltipDirection;

  /**
   * Whether the tooltip should be hidden.
   */
  @Prop() hideTooltip = false;

  /**
   * A pipe (`/`) seperated string of JS keyboard keys, that when caught in a `keydown` event, will
   * toggle the muted state of the player.
   */
  @Prop() muteKeys?: string = 'm';

  /**
   * Prevents the volume being changed using the Up/Down arrow keys.
   */
  @Prop() noKeyboard = false;

  @Watch('noKeyboard')
  onNoKeyboardChange() {
    this.keyboardDisposal.empty();
    if (this.noKeyboard) return;
    const player = findRootPlayer(this);
    this.keyboardDisposal.add(listen(player, 'keydown', (event: KeyboardEvent) => {
      if ((event.key !== 'ArrowUp') && (event.key !== 'ArrowDown')) return;
      const isUpArrow = (event.key === 'ArrowUp');
      const newVolume = isUpArrow ? Math.min(100, this.volume + 5) : Math.max(0, this.volume - 5);
      this.dispatch(PlayerProp.Volume, parseInt(`${newVolume}`, 10));
    }));
  }

  /**
   * @internal
   */
  @Prop() muted: PlayerProps[PlayerProp.Muted] = false;

  /**
   * @internal
   */
  @Prop() volume: PlayerProps[PlayerProp.Volume] = 50;

  @Watch('muted')
  @Watch('volume')
  onPlayerVolumeChange() {
    this.currentVolume = this.muted ? 0 : this.volume;

    if (!this.muted && this.prevMuted && this.volume === 0) {
      this.dispatch(PlayerProp.Volume, 30);
    }

    this.prevMuted = this.muted;
  }

  /**
   * @internal
   */
  @Prop() isMobile: PlayerProps[PlayerProp.IsMobile] = false;

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps[PlayerProp.I18N] = {};

  connectedCallback() {
    this.prevMuted = this.muted;
    this.dispatch = createPlayerStateDispatcher(this);
    this.onNoKeyboardChange();
  }

  disconnectedCallback() {
    this.keyboardDisposal.empty();
  }

  private onShowSlider() {
    clearTimeout(this.hideSliderTimeout);
    this.isSliderActive = true;
  }

  private onHideSlider() {
    this.hideSliderTimeout = setTimeout(() => {
      this.isSliderActive = false;
    }, 100) as any;
  }

  private onVolumeChange(event: CustomEvent<number>) {
    const newVolume = event.detail;
    this.currentVolume = newVolume;
    this.dispatch(PlayerProp.Volume, newVolume);
    this.dispatch(PlayerProp.Muted, newVolume === 0);
  }

  private onKeyDown(event: KeyboardEvent) {
    if ((event.key !== 'ArrowLeft') && (event.key !== 'ArrowRight')) return;
    event.stopPropagation();
  }

  render() {
    return (
      <Host
        onMouseEnter={this.onShowSlider.bind(this)}
        onMouseLeave={this.onHideSlider.bind(this)}
      >
        <vime-mute-control
          keys={this.muteKeys}
          lowVolumeIcon={this.lowVolumeIcon}
          highVolumeIcon={this.highVolumeIcon}
          mutedIcon={this.mutedIcon}
          tooltipDirection={this.tooltipDirection}
          hideTooltip={this.hideTooltip}
          onFocus={this.onShowSlider.bind(this)}
          onBlur={this.onHideSlider.bind(this)}
        />
        <vime-slider
          class={{
            hidden: this.isMobile,
            active: this.isSliderActive,
          }}
          step={5}
          max={100}
          value={this.currentVolume}
          label={this.i18n.volume}
          onKeyDown={this.onKeyDown.bind(this)}
          onFocus={this.onShowSlider.bind(this)}
          onBlur={this.onHideSlider.bind(this)}
          onVValueChange={this.onVolumeChange.bind(this)}
        />
      </Host>
    );
  }
}

openPlayerWormhole(VolumeControl, [
  PlayerProp.Volume,
  PlayerProp.Muted,
  PlayerProp.IsMobile,
  PlayerProp.I18N,
]);
