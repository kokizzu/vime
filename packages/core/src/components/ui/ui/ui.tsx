import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
