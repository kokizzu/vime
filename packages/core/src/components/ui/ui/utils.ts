import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
