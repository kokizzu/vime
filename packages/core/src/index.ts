import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
