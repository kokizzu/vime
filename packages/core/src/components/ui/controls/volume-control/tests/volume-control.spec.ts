import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../../ui/tests';
import { VolumeControl } from '../volume-control';

let page: SpecPage;
let control: HTMLVimeVolumeControlElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [VolumeControl],
    '<vime-volume-control />',
  ));

  control = page.root!.querySelector('vime-volume-control')!;
});

it('should be structurally sound', () => {
  expect(control).toMatchSnapshot();
});
