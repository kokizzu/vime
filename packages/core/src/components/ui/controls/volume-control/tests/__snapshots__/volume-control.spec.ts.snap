// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-volume-control>
  <vime-mute-control highvolumeicon="#vime-volume-high" keys="m" lowvolumeicon="#vime-volume-low" mutedicon="#vime-volume-mute"></vime-mute-control>
  <vime-slider label="Volume" max="100" step="5" value="50"></vime-slider>
</vime-volume-control>
`;
